### PR TITLE
[Error] Raise TaichiSyntaxError when multiple return path detected

### LINKED ISSUE #1840: [Lang] Confusing behavior when functions have multiple returns
**Describe the bug**
When using the `all(stmt)` judgment in taichi scope and the result is `False`, the function seems to stop unexpectedly. Please see the screenshots and code below.

**To Reproduce**
```python
import taichi as ti

@ti.func
def is_in_bound(vec: ti.template(), AABB_min: ti.template(), AABB_max: ti.template()):
    if all(AABB_min <= vec) and all(vec <= AABB_max):
        print('all judge is True')
        return 1
    else:
        print('all judge is False')
        return 0

@ti.kernel
def test():
    x_inside = ti.Vector([0.5, 0.5, 2.5])
    x_outside = ti.Vector([0.5, 0.5, 3.5])
    AABB_min = ti.Vector([0, 0, 0])
    AABB_max = ti.Vector([1, 2, 3])
    print('test inside point', is_in_bound(x_inside, AABB_min, AABB_max))
    print('---')
    print('test outside point', is_in_bound(x_outside, AABB_min, AABB_max))

test()
```
**Log**
image

### PR DESCRIPTION
**Pull request by archibate**
Related issue = fix #1840

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
```py
import taichi as ti

ti.init(print_preprocessed=True)

@ti.func
def sign(x):
    if x > 0:
        return 1
    else:
        return -1

@ti.kernel
def main():
    print(sign(-1))

main()
```
obtains:
```
(err-multiple-return) [bate@archit taichi]$ p a.py
[Taichi] mode=development
[Taichi] preparing sandbox at /tmp/taichi-8c3ueytb
[Taichi] <dev mode>, llvm 10.0.0, commit 09d665f9, linux, python 3.8.3
[Taichi] Starting on arch=x64
[Taichi] materializing...
Initial AST:
def main():
    print(sign(-1))

Preprocessed:
def main():
    import taichi as ti
    ti.ti_print(sign(-1))

Checked:
def main():
    import taichi as ti
    ti.ti_print(sign(-1))

Final AST:
def main():
    import taichi as ti
    ti.ti_print(sign(-1))

Initial AST:
def sign(x):
    if x > 0:
        return 1
    else:
        return -1

Preprocessed:
def sign(x_by_value__):
    import taichi as ti
    x = ti.expr_init_func(x_by_value__)
    if 1:
        __cond = ti.chain_compare([x, 0], ['Gt'])
        ti.begin_frontend_if(__cond)
        ti.core.begin_frontend_if_true()
        return 1
        ti.core.pop_scope()
        ti.core.begin_frontend_if_false()
        return -1
        ti.core.pop_scope()

Traceback (most recent call last):
  File "a.py", line 16, in <module>
    main()
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 574, in wrapped
    return primal(*args, **kwargs)
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 501, in __call__
    self.materialize(key=key, args=args, arg_features=arg_features)
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 370, in materialize
    taichi_kernel = taichi_kernel.define(taichi_ast_generator)
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 367, in taichi_ast_generator
    compiled()
  File "a.py", line 14, in main
    print(sign(-1))
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 42, in decorated
    return fun.__call__(*args)
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 86, in __call__
    self.do_compile()
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 98, in do_compile
    visitor.visit(tree)
  File "/home/bate/Develop/taichi/python/taichi/lang/transformer.py", line 49, in visit
    self.pass_Checks.visit(tree)
  File "/usr/lib/python3.8/ast.py", line 363, in visit
    return visitor(node)
  File "/usr/lib/python3.8/ast.py", line 439, in generic_visit
    value = self.visit(value)
  File "/usr/lib/python3.8/ast.py", line 363, in visit
    return visitor(node)
  File "/usr/lib/python3.8/ast.py", line 439, in generic_visit
    value = self.visit(value)
  File "/usr/lib/python3.8/ast.py", line 363, in visit
    return visitor(node)
  File "/home/bate/Develop/taichi/python/taichi/lang/transformer.py", line 904, in visit_If
    node.body = list(map(self.visit, node.body))
  File "/usr/lib/python3.8/ast.py", line 363, in visit
    return visitor(node)
  File "/home/bate/Develop/taichi/python/taichi/lang/transformer.py", line 919, in visit_Return
    raise TaichiSyntaxError(
taichi.lang.exception.TaichiSyntaxError: Taichi functions/kernels cannot have multiple returns! Consider use a local variable to walk around:

  def safe_sqrt(x):
    ret = 0.0
    if x >= 0:
      ret = ti.sqrt(x)
    else:
      ret = 0.0
    return ret
```

### CODE REVIEW

File: python/taichi/lang/transformer.py [hunk 1]
```diff
@@ -877,9 +877,58 @@ def visit_Return(self, node):
...
+                'Taichi functions/kernels cannot have multiple returns!'
+                ' Consider use a local variable to walk around:\n'
+                '''
+  def safe_sqrt(x):
+    ret = 0.0
+    if x >= 0:
+      ret = ti.sqrt(x)
+    else:
+      ret = 0.0
+    return ret
```
Comment (xumingkuan) [inline python/taichi/lang/transformer.py:931]: Users may wonder why their programs are related to `safe_sqrt`. Maybe remove this example and print sth like `Consider using a local variable to unify multiple returns.` instead?

File: python/taichi/lang/transformer.py [hunk 2]
```diff
@@ -877,9 +877,58 @@ def visit_Return(self, node):
 class ASTTransformerChecks(ASTTransformerBase):
     def __init__(self, func):
         super().__init__(func)
+        self.had_return = False
```
Comment (xumingkuan) [inline python/taichi/lang/transformer.py:880]: Maybe `has_return`? I don't get the point using the past tense.

**Review by xumingkuan — COMMENTED** [review]
Cool!

File: python/taichi/lang/transformer.py [hunk 3]
```diff
@@ -877,9 +877,49 @@ def visit_Return(self, node):
...
+        self.in_static_if = False
 
     def visit_Call(self, node):
         if isinstance(node.func, ast.Name):
             node.args = [node.func] + node.args
             node.func = self.parse_expr('ti.func_call_with_check')
         return node
+
+    @staticmethod
+    def get_taichi_decorator(node):
```
Comment (xumingkuan) [inline python/taichi/lang/transformer.py:890]: Let's simply reuse this here:
 https://github.com/taichi-dev/taichi/blob/64a612225d08a43d6ce907f196438e70f28c4679/python/taichi/lang/transformer.py#L347

File: python/taichi/lang/transformer.py [hunk 4]
```diff
@@ -877,9 +877,49 @@ def visit_Return(self, node):
...
+        if isinstance(node, ast.Call) and isinstance(
+                node.func, ast.Attribute) and isinstance(
+                    node.func.value, ast.Name) and node.func.value.id == 'ti':
+            attr = node.func.attr
+            if attr in ['static', 'grouped']:
+                return attr
+
+        return None
+
+    def visit_If(self, node):
```
Comment (xumingkuan) [inline python/taichi/lang/transformer.py:900]: Looks like we already had a `visit_If` in this file?

Comment (archibate) [reply python/taichi/lang/transformer.py:900]: We're moving towards multipass AST transforming system. So this is actually the second pass.

Comment (archibate) [reply python/taichi/lang/transformer.py:900]: > Edit: OIC, this is the second pass. Do they need to be in the second pass?

Not necessary, but putting things too tight together increases maintainance cost. The `ASTTransformerChecks` is a pass focus on sanity check. Composing sanity logics to already-complicated-preprocessor logics sounds too painful, isn't it?

**Comment by archibate** [comment]
The `Checks the Workflow Run of the Previous Commit` failed after 14m 36s:

> `urllib.error.HTTPError: HTTP Error 403: rate limit exceeded`

@k-ye is this expected? Seems we shouldn't request GitHub API so often.
Or we'd simply return success on `HTTPError`?

**Comment by k-ye** [comment]
> @k-ye is this expected? Seems we shouldn't request GitHub API so often.

Good catch! Sorry i ignored the authentication step. For unauthorized requests, Github had a limit of 60 requests/hour (ref: https://github.com/forestgeo/learn/issues/148#issuecomment-439581637). For **authorized, Github action** APIs, it is 1000 (https://docs.github.com/en/actions/getting-started-with-github-actions/about-github-actions#usage-limits). Note the number is somewhat different from authorized general APIs (which seems to be 5000)

> Or we'd simply return success on HTTPError?

No, let me make a proper fix following https://docs.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token

**Comment by xumingkuan** [comment]
I think this is not completely tested -- try this case:

```python3
@ti.test(arch=ti.cpu)
def test_func_multiple_return_in_if_and_static_if():
    @ti.func
    def safe_sqrt(a: ti.template()):
        if a > 0:
            if ti.static(a > 0):
                return ti.sqrt(a)
            else:
                return 0.0
        else:
            return 0.0

    @ti.kernel
    def kern():
        print(safe_sqrt(-233))

    with pytest.raises(ti.TaichiSyntaxError,
                       match='cannot have multiple returns'):
        kern()
```

**Comment by archibate** [comment]
> I think this is not completely tested -- try this case:
> 
> ```python
> @ti.test(arch=ti.cpu)
> def test_func_multiple_return_in_if_and_static_if():
>     @ti.func
>     def safe_sqrt(a: ti.template()):
>         if a > 0:
>             if ti.static(a > 0):
>                 return ti.sqrt(a)
>             else:
>                 return 0.0
>         else:
>             return 0.0
> 
>     @ti.kernel
>     def kern():
>         print(safe_sqrt(-233))
> 
>     with pytest.raises(ti.TaichiSyntaxError,
>                        match='cannot have multiple returns'):
>         kern()
> ```

This seems very hard to test to me.
For example, this is ERROR:
```py
     @ti.func
     def safe_sqrt(a: ti.template()):
         if a > 0:
             if ti.static(a > 2):
                 return ti.sqrt(a)
             else:
                 return 0.0
         else:
             return 0.0
```
but this is OK:
```py
     @ti.func
     def safe_sqrt(a: ti.template()):
         if a > 0:
             if ti.static(a > 2):
                 return ti.sqrt(a)
             elif ti.static(a > 1):
                 return 0.0
         else:
             return 0.0
```

 I think it's actually alright. If one is using static-if, they're likely professional and know what they're doing.

File: python/taichi/lang/transformer.py [hunk 5]
```diff
@@ -877,9 +875,38 @@ def visit_Return(self, node):
...
+    def visit_Return(self, node):
+        if self.in_static_if:  # we can have multiple return in static-if branches
+            return node
+
+        if not self.has_return:
+            self.has_return = True
+        else:
+            raise TaichiSyntaxError(
+                'Taichi functions/kernels cannot have multiple returns!'
+                ' Consider use a local variable to walk around')
```
Comment (xumingkuan) [inline python/taichi/lang/transformer.py:910]:
```suggestion
                ' Consider using a local variable to walk around.')
```

**Review by xumingkuan — APPROVED** [review]
I see. LGTM now with a nit :)

**Pull request merged**